scripts/release/other_identifiers/retrieve_indirectIdentifiers_from_mart.pl

```diff
@@ -1,77 +1,156 @@
-     -pass   => $opt_pass,
-     -port   => $opt_port,
-     -dbname => $opt_db,
-#     -DEBUG => $opt_debug
-     );
-
-# Fetch the species
-$opt_sp ||= 'Homo sapiens';
-
-my $sp = $dba->fetch_instance_by_attribute('Species',[['name',[$opt_sp]]])->[0]
-    || $logger->error_die("No species '$opt_sp' found.\n");
-my $sp_mart_name;
-if ($sp->displayName =~ /^(\w)\w+ (\w+)$/) {
-    $sp_mart_name = lc("$1$2");
-} else {
-    $logger->error_die("Can't form species abbreviation for mart from '" . $sp->displayName . "'.\n");
+
+run(@ARGV) unless caller();
+
+sub run {
+    Log::Log4perl->init(\$LOG_CONF);
+    my $logger = get_logger(__PACKAGE__);
+
+    our($species_code, $help);
+
+    &GetOptions(
+        "sp=s" => \$species_code,
+        "help" => \$help
+    );
+
+    if ($help) {
+        print usage_instructions();
+        exit;
+    }
+
+    foreach my $species_abbreviation (get_species_to_query($species_code)) {
+        my $species_mart_name = get_species_mart_name($species_abbreviation);
+
+        foreach my $identifier (get_identifiers($species_mart_name)) {
+            next if $identifier =~ /chembl|clone_based|dbass|description|ottg|ottt|ottp|shares_cds|merops|mirbase|reactome/;
+
+            my $output_file = "output/$species_mart_name\_$identifier";
+            if (-s $output_file) {
+                $logger->info("$output_file already exists - creating backup");
+                system("mv -f $output_file $output_file.bak");
+            }
+            download_identifier_file($species_abbreviation, $identifier, $output_file);
+        }
+    }
+}
+
+# Returns all configured species abbreviations by default
+# or the single species abbreviation if it exists in the
+# @species array imported from Config_Species.pm
+sub get_species_to_query {
+    my $selected_species = shift;
+
+    my $logger = get_logger(__PACKAGE__);
+
+    my @species_to_query;
+    if ($selected_species) {
+        if (any { $_ eq $selected_species } @species) {
+            @species_to_query = ($selected_species);
+        } else {
+            $logger->logconfess("$selected_species species: EnsEMBL mart information is unknown");
+        }
+    } else {
+        @species_to_query = @species;
+    }
+
+    return @species_to_query;
+}
+
+# Retrieves the appropriate wget query for the species and
+# external identifier and then attemps to download the
+# identifier file and save it to an output file.
+# The function will recurse on failure to re-try
+# downloading to a pre-set maximum attempt number
+sub download_identifier_file {
+    my $species_abbreviation = shift;
+    my $identifier = shift;
+    my $output_file = shift;
+    my $attempt_count = shift // 1;
+    my $max_attempts = 3;
+
+    my $logger = get_logger(__PACKAGE__);
+
+    my $results = run_wget_query($species_abbreviation, $identifier);
+
+    my $output = $results->{'output'};
+    my $completion_tag = qr/\[success\]\n/;
+    if ($output =~ /$completion_tag/) {

```

Comment:
I never knew how that worked -- when testing biomart orthopairs in Java it always had the pesky success message